controllers/toolchaincluster_resources/toolchaincluster_resources_controller.go

```diff
@@ -0,0 +1,77 @@
+package toolchaincluster_resources
+
+import (
+	"context"
+	"embed"
+	"fmt"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/api/v1alpha1"
+	commoncontroller "github.com/codeready-toolchain/toolchain-common/controllers"
+	applycl "github.com/codeready-toolchain/toolchain-common/pkg/client"
+	commonpredicates "github.com/codeready-toolchain/toolchain-common/pkg/predicate"
+	"github.com/codeready-toolchain/toolchain-common/pkg/template"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// ResourceControllerLabelValue is being added to all the resources managed by this controller.
+// It's then used to filter all the events on those resources by using a mapper function in the watcher configuration.
+const ResourceControllerLabelValue = "toolchaincluster-resources-controller" // todo move this label value to api repo
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager, operatorNamespace string) error {
+	// check for required templates FS directory
+	if r.templates == nil {
+		return fmt.Errorf("no templates FS configured")
+	}
+
+	build := ctrl.NewControllerManagedBy(mgr).
+		For(&v1.ServiceAccount{})
+
+	// add watcher for all kinds from given templates
+	var err error
+	r.templateObjects, err = template.LoadObjectsFromEmbedFS(r.templates, &template.Variables{Namespace: operatorNamespace})
+	if err != nil {
+		return err
+	}
+	mapToOwnerByLabel := handler.EnqueueRequestsFromMapFunc(commoncontroller.MapToControllerByMatchingLabel(toolchainv1alpha1.ProviderLabelKey, ResourceControllerLabelValue))
+	for _, obj := range r.templateObjects {
```

Comment:
just one idea - what if the template contains more objects of the same kind? wouldn't the builder complain that you are trying to watch the same kind multiple times?

Reply:
Good point!

TBH I'm not sure what will happen in this case , by quickly looking at the builder code I don't think it will error, but maybe worth adding an e2e test case for this ?

Reply:
I don't think that we need an e2e for this. After all, if we change the template and it would be a problem for the controller/builder, then it would start failing and the current e2e tests would fail. 
It was just thought, if the builder seems to be ok if we call it more times with the same kind, then keep it as it is :wink:

Reply:
Right, it should start failing right away with the existing e2e tests, good point!